### PR TITLE
WIP Hacking stuff

### PR DESCRIPTION
**Pull request by nphyx**
#### Summary
```SUMMARY: Content "adds a bunch of fun hacky stuff for computers and electronics"```

#### Purpose of change
This is more of a long-term project. I'm just adding it here for feedback / ideas / commentary.

The general concept is to add a bunch of stuff that you could *realistically* do with enough time and skill even after the end of the world, to generally make electronics clutter more interesting, and to flesh out robot creation and interaction therewith.

I don't expect to have to make a lot of code changes for this. Most of the infrastructure is already there, but dormant. It just needs a lot of JSON.

#### Describe the solution
Make a bunch of new stuff for people to do with computers and electronics.

#### Describe alternatives you've considered
Waiting until the world ends IRL to live out my dream of building killer junkbots out of common household electronics.

#### Testing
None yet, this is totally unusable. Don't try it out. I will test as I go along.

#### Additional context
I'll be trying to stay within the themes and bounds of the game so this can make it in as core content. To that end I'd appreciate comments, guidance, and "no way that's going to work, don't bother".

**Design notes (WIP)**:
This draft is just brainstorming/notes/screwing with things to see what works. It will be broken into separate PRs.

Assumptions about game world:
- AI cores use specialized processors/ASICs/SoCs, probably along the lines of https://www.allaboutcircuits.com/news/ai-hardware-roundup-neuromorphic-chips
  - you'd need to salvage these in order to reprogram them and/or build your own cores
    - or you could use a standard processor for really low-end/glitchy/badly performing AI cores
  - they use a standard serial bus so you could wire them to other electronic parts, rather than the whole system being specialized and non-reproducible

Gameplay intent:
- looting/killing/salvaging stuff should be generally the faster way to get components you need, but making them with electronics & computers is an alternate route if RNG isn't on your side or you just want to for fun (as with fabrication & mechanics)
- start as a hacker, student, or technician has a similar feel to starting as a mechanic or survivalist - a slow burn and you depend on your gadgets to carry you in high-risk situations, but it pays off later if you invest in it

Things you should be able to do:
- make your own circuits - either by directly soldering components, using a breadboard, or on a handmade PCB
  - for PCB you'll need a blank, strong acid, and some acid-resistant ink either from a marker or inkjet printer, but this can be done without any specialized equipment. these parts would be found in labs and electronic hobby/repair shops
- wire together pre-existing ICs and electronic components out of salvage to produce novel gadgets (mid-level electronic skill)
  - janky and inefficient, but if you have reference material or don't mind letting the smoke out of a lot of parts to figure out how they work, you can do it
  - working with ICs and SMDs (abstracted as electronic scrap) would require a professional a soldering station, not a funky handmade iron - could be found in labs, electronic shops, etc. 
- write and install software that does equivalent of existing gadgets (mp3 player, eink tablet, etc) with mid to high level computer skill and lots of time
  - this already works in-game by making software a toolmod that adds qualities/actions/flags
  - programming tasks would be really long-duration things you'd work on in spare time over the course of a season or two
    - the new crafting system makes this feasible
- build your own custom bots & drones (as monster types)
  - already possible for hacks, spiders, etc.
  - they wouldn't be equal to milspec bots, but good for trashing zeds
  - would be cool to build drones with cameras for scouting, but this doesn't seem to be supported (separate PR later if so)
  - or arm them, board yourself up in a bathroom, and go kill stuff!
    - Ouch! something hurts, do you want to stop piloting your drone?
- larger/more complex drones using vehicles system
   - sort of already possible, but could be expanded
- build and program a 3D printer (ala RepRap)
    - used for printing more complex plastic bits than you can make with a mold (plastic robot limbs, etc)
   - need to examine how set-and-forget objects work (charcoal kiln, smoker, etc)
    - you'll need to find or make an extruder to produce filament from plastic.
      - IRL these really suck compared to commercial models (see https://www.filabot.com/)
      - but can be bootstrapped from common parts if you have some filament and a working printer

Things you can't do:
- make your own ICs - this requires specialized equipment, a cleanroom, and a hell of a lot of design work (maybe a nanofab could produce some pre-existing parts using a template)
- build anything comparable to commercial devices from scratch/with random parts
  - but you might be able to make something like a SBC using smartphone parts
  - or a really bulky, fragile custom portable computer from misc. computer parts
- build milspec bots from scratch
- build novel CBMs
  - but salvaging & repairing them might be easier with craft chains that supply some of the needed tools

Complexity:
- probably not justified to add new islots for electronic gadgets or model disk space/memory/processing power/etc
- instead, going to try to make software a specialized toolmod that works with compatible devices (current implementation is a hack and won't be kept, it doesn't even work right - instead needs a new actor to "install software", but this will be straightforward to write)

### CODE REVIEW

File: data/json/items/software.json [hunk 1]
```diff
@@ -80,5 +88,67 @@
     "name": { "str": "integrated circuit datasheet archives", "str_pl": "misc software" },
     "description": "Huge archives of numerous IC circuit datasheets from several major manufacturers.  Probably valuable to the right person, as it would be hard to salvage and reuse these components without them.",
     "price": 40000
+  },
+  {
+    "abstract": "mod_software",
+    "type": "TOOLMOD",
+    "container": "usb_drive",
+    "name": { "str": "USB with usable software" },
```
Comment (anothersimulacrum) [inline data/json/items/software.json:96]:
```suggestion
    "name": { "str": "USB with usable software", "str_pl": "USBs with usable software" },
```

File: data/json/items/software.json [hunk 2]
```diff
@@ -80,5 +88,67 @@
...
+    "material": [ "plastic" ],
+    "color": "dark_gray",
+    "symbol": ","
+  },
+  {
+    "id": "software_compiler_bootstrap",
+    "copy-from": "mod_software",
+    "type": "TOOLMOD",
+    "flags": [ "SOFTWARE" ],
+    "name": { "str": "bootstrap compiler", "str_pl": "bootstrap compilers" },
```
Comment (anothersimulacrum) [inline data/json/items/software.json:106]:
```suggestion
    "name": { "str": "bootstrap compiler" },
```

File: data/json/items/software.json [hunk 3]
```diff
@@ -80,5 +88,67 @@
...
+    "description": "A very basic compiler written directly in binary machine code by some mad genius. It's not much good for anything besides compiling a better compiler on a machine that has no development tools, but you could install it on a computer.",
+    "qualities": [ [ "SOFTWARE_COMPILER", 1 ] ]
+  },
+  {
+    "id": "software_compiler_homebrew",
+    "copy-from": "mod_software",
+    "type": "TOOLMOD",
+    "flags": [ "SOFTWARE" ],
+    "name": { "str": "homebrew compiler", "str_pl": "homebrew compilers" },
```
Comment (anothersimulacrum) [inline data/json/items/software.json:115]:
```suggestion
    "name": { "str": "homebrew compiler" },
```

File: data/json/items/software.json [hunk 4]
```diff
@@ -80,5 +88,67 @@
...
+    "description": "A competently written compiler for a homebrew programming language, along with documentation for the language itself and a few useful libraries. It doesn't do any fancy optimization and lacks cross-platform compatibility, but you can write real code with it - albeit with a very simplistic language.",
+    "qualities": [ [ "SOFTWARE_COMPILER", 1 ] ]
+  },
+  {
+    "id": "software_decompiler",
+    "copy-from": "mod_software",
+    "type": "TOOLMOD",
+    "flags": [ "SOFTWARE" ],
+    "name": { "str": "decompiler", "str_pl": "decompilers" },
```
Comment (anothersimulacrum) [inline data/json/items/software.json:124]:
```suggestion
    "name": { "str": "decompiler" },
```

File: data/json/items/software.json [hunk 5]
```diff
@@ -80,5 +88,67 @@
...
+    "description": "A hex editor, decompiler and some associated analytic software. If you were in an apocalyptic scenario and had with no access to code documentation, you might use it to gain some insight into the operation and APIs of some compiled code. It would be a huge mind-numbing pain, but it would be possible.",
+    "qualities": [ [ "SOFTWARE_DECOMPILER", 1 ] ]
+  },
+  {
+    "id": "software_compiler_professional",
+    "copy-from": "mod_software",
+    "type": "TOOLMOD",
+    "flags": [ "SOFTWARE" ],
+    "name": { "str": "professional compiler", "str_pl": "professional compilers" },
```
Comment (anothersimulacrum) [inline data/json/items/software.json:133]:
```suggestion
    "name": { "str": "professional compiler" },
```

File: data/json/items/software.json [hunk 6]
```diff
@@ -80,5 +88,67 @@
...
+    "description": "A suite of professional software development tools, including an IDE, compiler, build system, and an archive of useful software libraries with offline documentation.  You could install it in a computer and write some code.",
+    "qualities": [ [ "SOFTWARE_COMPILER", 3 ] ]
+  },
+  {
+    "id": "software_ai_sdk_makeshift",
+    "copy-from": "mod_software",
+    "type": "TOOLMOD",
+    "flags": [ "SOFTWARE" ],
+    "name": { "str": "makeshift AI SDK", "str_pl": "makeshift AI SDK" },
```
Comment (anothersimulacrum) [inline data/json/items/software.json:142]: Doesn't SDKs work?
```suggestion
    "name": { "str": "makeshift AI SDK" },
```

**Comment by nphyx** [comment]
Updated OP with design notes. Thanks @anothersimulacrum for the fixes :)

**Comment by crazykid080** [comment]
This sounds like a very interesting concept! I look forward to seeing this be fleshed out and built on!

**Comment by FirebrandCoding** [comment]
...you've effectively obsoleted a big part of [Homestead](https://github.com/FirebrandCoding/HomesteadInstrumentsInc).

Well done, but also... Crap.

**Comment by crazykid080** [comment]
One interesting idea I had (maybe a minute ago) would be LANs for more important buildings. for example, a lab would have secure networks to prevent hackers from accessing more important areas protected by defensive modules, no idea if this is even possible so feel free to tweak or just ignore this one

**Comment by nphyx** [comment]
> ...you've effectively obsoleted a big part of [Homestead](https://github.com/FirebrandCoding/HomesteadInstrumentsInc).
> 
> Well done, but also... Crap.

Nah I checked it out and it looks like you're aiming at setting up a full scale fablab, which is way outside the scope of what I want to do.

For the printer I'm thinking more like a jank handmade reprap made from robot salvage and parts from a hardware store. I don't know if you've ever used one of those before but they ... leave a lot to be desired compared to something ordered from a professional print shop like shapeways.

It would be sufficient to make a handful of things that are actually useful, but I'm thinking of home-printed variants that would go into producing a lower-quality makeshift version of whatever you're trying to replicate.

You should be able to find a better one as rare loot from a lab or research center.

I am working on a PR with the first couple prerequisites right now. Among other things proper code support for installing software on computers, which should be useful to you. It'll add a new `use_action`, "install software", that you can tag onto your working computers, and a new flag for toolmods, `SOFTWARE`, that you put on installable software. It also lifts any `use_action`s from toolmods into the activate menu for the tool they're installed on.

Or at least that's the plan, not sure how it will look if it makes it through review. It's mostly working already but buggy - should be in good shape sometime this week depending on how much time I get to work on it.

Check out the sketches in this PR for what I had in mind for how programming will work too.

> One interesting idea I had (maybe a minute ago) would be LANs for more important buildings. for example, a lab would have secure networks to prevent hackers from accessing more important areas protected by defensive modules, no idea if this is even possible so feel free to tweak or just ignore this one

That would be really cool but I don't know how to even start on that yet. Long-term I'm hoping the changes I introduce here will be enough to create a cyberpunk-themed mod. The base game already has bionics ofc which is a big first step, but I want more robust support for decking & rigging.

**Comment by FirebrandCoding** [comment]
> which is way outside the scope of what I want to do

Oh, I didn't say you've obsoleted the whole mod – but I *did* work hard on bringing working computers into the game, and you've just... swept that up and did something useful with it. If you would implement servers as a measure of distributed computation for large projects – like full AI design modelling – I'd just delete that part entirely and focus on production and energy needs.

> For the printer I'm thinking more like a jank handmade reprap made from robot salvage and parts from a hardware store.

That's not the idea. Printers *are* meant to be professional, store-quality tools. I'm going to implement assembly kits for each of them, and both the kits and the assembled printers – among other things – could be found in stores and at homes, 'cause you know: that's real life.

Creating a home-made printer should be a good idea, though. It sure is possible at the moment, even though the quality is to be desired.

> Among other things proper code support for installing software on computers, which should be useful to you.

If you do that, I'll just throw that part of the mod away, because there's no longer a need! You've already drafted most of what I wanted from computational capacities at this stage.

**Comment by nphyx** [comment]
> That's not the idea. Printers _are_ meant to be professional, store-quality tools. I'm going to implement assembly kits for each of them, and both the kits and the assembled printers – among other things – could be found in stores and at homes, 'cause you know: that's real life.

That's what I mean - I don't think what I'm doing here will step on your toes, your printers will probably be much higher quality. But I'll look at yours and see if I can make them inter-operable.

> Regarding LAN: the idea I had was to search for computers in the vicinity (the standard crafting radius is 6 tiles, which would suffice for PC-to-server connection for distributed computation), _or_ be able to connect them with a cable a la Ethernet. You might be able to code the latter for some automated cable layouts.

Possibly, I haven't begun to investigate that yet and not sure I need it, but I would like to see it so I might tackle it someday.

The direction I'm going with this PR is more "lone crazy hacker making impractical stuff out of salvage", less "let's reboot civilization!".

Your mod reminds me of Open Source Ecology or *Build Your Own Metal Working Shop from Scrap*, which is very cool.

**Comment by FirebrandCoding** [comment]
> But I'll look at yours and see if I can make them inter-operable.

Let me know if I need to make some changes, either. I'd rather align my mod with the base game and expand on it than try and make something completely different.

> The direction I'm going with this PR is more "lone crazy hacker making impractical stuff out of salvage", less "let's reboot civilization!".

Sure thing. I'm thinking, though: at some point you either might have to or could afford to go a more practical route. 3D-printing RAM boards and soldering salvaged memory chips onto it could be fun if your survivor is that level of competent. :)

Either way: looking forward to see how this develops! The premise sounds awesome, and what you've done so far looks interesting already.

**Comment by nphyx** [comment]
this draft is superseded by individual prs.

**Pull request closed without merging**